### PR TITLE
Softdels part 1 (Datum softdels)

### PR DESCRIPTION
**Pull request by moonheart08**
Implements softdel support for /datum, changing internal refs to be weak as to avoid pinning the objects forever.

### CODE REVIEW

File: OpenDreamRuntime/Objects/Types/DreamObjectMovable.cs [hunk 1]
```diff
@@ -198,4 +204,5 @@ public class DreamObjectMovable : DreamObjectAtom {
                 throw new ArgumentException($"Invalid loc {loc}");
         }
     }
+
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectMovable.cs:207]:
```suggestion
```

File: OpenDreamRuntime/Objects/Types/DreamObjectSavefile.cs [hunk 2]
```diff
@@ -124,9 +124,22 @@ public sealed class DreamObjectSavefile : DreamObject {
...
+        if (possiblyThreaded) {
+            EnterIntoDelQueue();
+            return;
+        }
+
+        if (Deleted)
+            return;
+
+        if (Resource is null)
+            return; // Seemingly a long-standng issue, I don't know how to fix this, and it only now appears due to the fact objects always Del() now.
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectSavefile.cs:138]:
```suggestion
            return; // Seemingly a long-standing issue, I don't know how to fix this, and it only now appears due to the fact objects always Del() now.
```

File: OpenDreamRuntime/Objects/Types/DreamObjectSavefile.cs [hunk 3]
```diff
@@ -124,9 +124,22 @@ public sealed class DreamObjectSavefile : DreamObject {
...
-    protected override void HandleDeletion() {
+    protected override void HandleDeletion(bool possiblyThreaded) {
+        // SAFETY: Close() is not threadsafe and doesn't have reason to be.
+        if (possiblyThreaded) {
+            EnterIntoDelQueue();
+            return;
+        }
+
+        if (Deleted)
+            return;
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/Types/DreamObjectSavefile.cs:135]: `DreamObject.Delete()` should already be checking this before calling `HandleDeletion()`

File: OpenDreamRuntime/DreamManager.cs [hunk 4]
```diff
@@ -203,10 +214,15 @@ public sealed partial class DreamManager {
...
                     }
-                    if (!ReferenceIDs.TryGetValue(refObject, out idx)) {
-                        idx = _dreamObjectRefIdCounter++;
-                        ReferenceIDs.Add(refObject, idx);
-                        ReferenceIDsToDreamObject.Add(idx, refObject);
+                    if (refObject.RefID is not {} id) {
+                        idx = Interlocked.Increment(ref _dreamObjectRefIdCounter);
+                        refObject.RefID = idx;
+
+                        // SAFETY: Infalliable! idx is always unique and add can only fail if this is not the case.
```
Comment (wixoaGit) [inline OpenDreamRuntime/DreamManager.cs:221]:
```suggestion
                        // SAFETY: Infallible! idx is always unique and add can only fail if this is not the case.
```

File: OpenDreamRuntime/DreamManager.cs [hunk 5]
```diff
@@ -203,10 +214,15 @@ public sealed partial class DreamManager {
...
+                        ReferenceIDsToDreamObject.TryAdd(idx, new WeakDreamRef(refObject));
+
```
Comment (wixoaGit) [inline OpenDreamRuntime/DreamManager.cs:223]:
```suggestion
```

File: OpenDreamRuntime/Objects/DreamObject.cs [hunk 6]
```diff
@@ -106,14 +109,41 @@ public class DreamObject {
             ObjectDefinition = null!;
         }
 
-        public void Delete() {
+
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/DreamObject.cs:112]:
```suggestion
```

File: OpenDreamRuntime/Objects/DreamObject.cs [hunk 7]
```diff
@@ -106,14 +109,41 @@ public class DreamObject {
...
+
+        /// <summary>
+        ///     Enters the current dream object into a global del queue that is guaranteed to run on the DM thread.
+        ///     Use if your deletion handler must be on the DM thread.
+        /// </summary>
+        protected void EnterIntoDelQueue() {
+            DreamManager.DelQueue.Add(this);
+        }
+        /// <summary>
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/DreamObject.cs:121]:
```suggestion
        }
        
        /// <summary>
```

File: OpenDreamRuntime/Objects/DreamObject.cs [hunk 8]
```diff
@@ -106,14 +109,41 @@ public class DreamObject {
...
+        public void Delete(bool possiblyThreaded = false) {
             if (Deleted)
                 return;
 
-            if (TryGetProc("Del", out var delProc))
-                DreamThread.Run(delProc, this, null);
+            if (TryGetProc("Del", out var delProc)) {
+                // SAFETY: See associated comment in Datum.dm. This relies on the invariant that this proc is in a
+                //         thread-safe subset of DM (if such a thing exists) or empty. Currently, it is empty.
+                var datumBaseProc = delProc.OwningType.Path == "/datum";
```
Comment (wixoaGit) [inline OpenDreamRuntime/Objects/DreamObject.cs:132]:
```suggestion
                var datumBaseProc = delProc.OwningType == ObjectTree.Datum;
```

Overriding `/datum/Del()` would still lead this to be true. Maybe also check if it's a DMProc with an empty Bytecode?

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamRuntime/Objects/DreamObject.cs (hunk 8 above)
Comment (moonheart08) [reply OpenDreamRuntime/Objects/DreamObject.cs:132]: I should probably just add a cached "safe to skip calling" flag to procs, as checking here would be bad for perf.

**Comment by wixoaGit** [comment]
Closing in favor of #1920

**Pull request closed without merging**